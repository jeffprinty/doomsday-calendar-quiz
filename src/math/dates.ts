import dayjs, { Dayjs } from 'dayjs';
import dayOfYear from 'dayjs/plugin/dayOfYear';
import duration from 'dayjs/plugin/duration';
import isLeapYear from 'dayjs/plugin/isLeapYear';
import isoWeek from 'dayjs/plugin/isoWeek';
import localeData from 'dayjs/plugin/localeData';
import localizedFormat from 'dayjs/plugin/localizedFormat';
import objectSupport from 'dayjs/plugin/objectSupport';
import timezone from 'dayjs/plugin/timezone';

import 'dayjs/locale/en';
import 'dayjs/locale/fr';
import 'dayjs/locale/en-gb';

import randomInteger from 'random-int';

import { getMnemonicForMonth, Mnemonic } from './month-doomsdays';
import { monthNames } from './months';
import { Weekday } from './weekdays';

dayjs.extend(objectSupport);
dayjs.extend(isLeapYear);
dayjs.extend(timezone);
dayjs.extend(isoWeek);
dayjs.extend(dayOfYear);
dayjs.extend(duration);
dayjs.extend(localizedFormat);
dayjs.extend(localeData);

dayjs.tz.guess();

export const formatFullDateWithWeekday = (date: Dayjs) => date.format('dddd LL');

export const getFirstDateForCalendar = (year: number) => {
  return dayjs({ year, month: 0, date: 1 }).startOf('week');
};

export const getDateFromMonthMnemonic = (monthMnemonic: Mnemonic, year: number): Dayjs => {
  const { common, leap, monthNumber } = monthMnemonic;
  const commonDate = dayjs(`${year}-${monthNumber}-${common}`);
  if ([1, 2].includes(monthNumber) && commonDate.isLeapYear()) {
    return dayjs(`${year}-${monthNumber}-${leap}`);
  }
  return commonDate;
};

export const dateIsLeapYear = (date: Dayjs) => date.isLeapYear();

export const getFullWeekday = (date: Dayjs) => dayjs(date).format('dddd');
export const getWeekdayForDate = (date: Dayjs): Weekday => dayjs(date).format('ddd') as Weekday;

export const getDoomsdayForYear = (year: number): Dayjs => dayjs({ year, month: 3, date: 4 });
export const getDoomsdayWeekdayForYear = (year: number): Weekday =>
  getWeekdayForDate(getDoomsdayForYear(year));
export const getDoomsdayFullWeekdayForYear = (year: number) =>
  getFullWeekday(getDoomsdayForYear(year));

export const getRandomDateInYear = (year: number) => {
  const randomDayOfYear = randomInteger(1, 365);
  return dayjs(`${year}-01-01`).dayOfYear(randomDayOfYear);
};

export const getDoomsdayWithinMonth = (date: Dayjs, common: number, leap?: number) =>
  dayjs({
    year: date.year(),
    month: date.month() + 1,
    date: leap || common,
  });

export const getRandomDateInModernity = () => {
  const randomYear = randomInteger(1900, 2099);
  return getRandomDateInYear(randomYear);
};

export const formatDayjsGuessDate = (date: Dayjs) => dayjs(date).format('LL');

// i18n
export const formatYearlessDate = (date: Dayjs) =>
  dayjs(date)
    .format('l')
    .replace(/\/\d{4}\b/, '');
export const formatYearlessDateShortMonth = (date: Dayjs) =>
  dayjs(date)
    .format('ll')
    .replace(/, \d{4}\b/, '');

export const getMonthMnemonicForDate = (date: Dayjs): Mnemonic => {
  const monthName = monthNames[date.month()];
  return getMnemonicForMonth(monthName);
};
