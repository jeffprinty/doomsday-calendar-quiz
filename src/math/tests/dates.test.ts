import { expect, test } from 'vitest';

import { getDoomsdayForYearNew } from '../dates';
import { doomsyearTable } from '../year';

for (const [year] of doomsyearTable) {
  test(`tests getDoomsdayForYearNew ${year}`, () => {
    expect(getDoomsdayForYearNew(year)).toMatchSnapshot();
    // expect(getDoomsdayForYearNew(year)).toEqual(getDoomsdayForYear(year))
  });
}
