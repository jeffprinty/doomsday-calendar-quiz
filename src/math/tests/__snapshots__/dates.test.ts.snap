// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tests getDoomsdayForYearNew 2000 1`] = `"2000-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2001 1`] = `"2001-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2002 1`] = `"2002-04-04T06:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2003 1`] = `"2003-04-04T06:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2004 1`] = `"2004-04-04T06:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2005 1`] = `"2005-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2006 1`] = `"2006-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2007 1`] = `"2007-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2008 1`] = `"2008-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2009 1`] = `"2009-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2010 1`] = `"2010-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2011 1`] = `"2011-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2012 1`] = `"2012-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2013 1`] = `"2013-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2014 1`] = `"2014-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2015 1`] = `"2015-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2016 1`] = `"2016-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2017 1`] = `"2017-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2018 1`] = `"2018-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2019 1`] = `"2019-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2020 1`] = `"2020-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2021 1`] = `"2021-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2022 1`] = `"2022-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2023 1`] = `"2023-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2024 1`] = `"2024-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2025 1`] = `"2025-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2026 1`] = `"2026-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2027 1`] = `"2027-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2028 1`] = `"2028-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2029 1`] = `"2029-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2030 1`] = `"2030-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2031 1`] = `"2031-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2032 1`] = `"2032-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2033 1`] = `"2033-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2034 1`] = `"2034-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2035 1`] = `"2035-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2036 1`] = `"2036-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2037 1`] = `"2037-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2038 1`] = `"2038-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2039 1`] = `"2039-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2040 1`] = `"2040-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2041 1`] = `"2041-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2042 1`] = `"2042-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2043 1`] = `"2043-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2044 1`] = `"2044-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2045 1`] = `"2045-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2046 1`] = `"2046-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2047 1`] = `"2047-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2048 1`] = `"2048-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2049 1`] = `"2049-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2050 1`] = `"2050-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2051 1`] = `"2051-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2052 1`] = `"2052-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2053 1`] = `"2053-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2054 1`] = `"2054-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2055 1`] = `"2055-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2056 1`] = `"2056-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2057 1`] = `"2057-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2058 1`] = `"2058-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2059 1`] = `"2059-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2060 1`] = `"2060-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2061 1`] = `"2061-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2062 1`] = `"2062-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2063 1`] = `"2063-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2064 1`] = `"2064-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2065 1`] = `"2065-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2066 1`] = `"2066-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2067 1`] = `"2067-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2068 1`] = `"2068-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2069 1`] = `"2069-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2070 1`] = `"2070-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2071 1`] = `"2071-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2072 1`] = `"2072-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2073 1`] = `"2073-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2074 1`] = `"2074-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2075 1`] = `"2075-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2076 1`] = `"2076-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2077 1`] = `"2077-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2078 1`] = `"2078-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2079 1`] = `"2079-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2080 1`] = `"2080-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2081 1`] = `"2081-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2082 1`] = `"2082-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2083 1`] = `"2083-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2084 1`] = `"2084-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2085 1`] = `"2085-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2086 1`] = `"2086-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2087 1`] = `"2087-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2088 1`] = `"2088-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2089 1`] = `"2089-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2090 1`] = `"2090-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2091 1`] = `"2091-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2092 1`] = `"2092-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2093 1`] = `"2093-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2094 1`] = `"2094-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2095 1`] = `"2095-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2096 1`] = `"2096-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2097 1`] = `"2097-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2098 1`] = `"2098-04-04T05:00:00.000Z"`;

exports[`tests getDoomsdayForYearNew 2099 1`] = `"2099-04-04T05:00:00.000Z"`;
